### ISSUE TITLE
OKHttpClient - recommended use of the client as static/single instance

### ISSUE THREAD

**Issue by updatede** [mendhak/gpslogger#687]
"OkHttpClients should be shared"
"OkHttp performs best when you create a single OkHttpClient instance and reuse it for all of your HTTP calls. This is because each client holds its own connection pool and thread pools. Reusing connections and threads reduces latency and saves memory. Conversely, creating a client for each request wastes resources on idle pools."
Above is quoted from "https://square.github.io/okhttp/3.x/okhttp/okhttp3/OkHttpClient.html"
But I see you create okhttpclient for every custom url log job, produce many garbage connections to server and wastes memory of mobile device.

**Comment by mendhak** [comment]
I can try adding some of the shutdown calls towards the end, since that matches up with the job queue scenario. 

![image](https://user-images.githubusercontent.com/746276/47280512-9c79b000-d5ce-11e8-8858-351aab48884d.png)

**Comment by updatede** [comment]
shutdown is unnecessary and it still waste memory,connection, cpu, thread,battery. it also add latency, because new connection add  tls handshakes and round trips. it waste server resource too. so, reuse connection is the right thing should be done as okhttp doc said. finally I decide to setup development environment and build it myself. I create instance of  okhttpclient and requestbuilder only once at  initiialization as static member of customurljob. so every instance of customurljob share one instance of okhttpclient. it works perfectly, garbage connections disappear , just one connection.

**Comment by mendhak** [comment]
That sounds like good news - if you were able to get the client as a static in the custom URL job there is hope that this can be done.

However there are multiple constraints here: 

* The custom URL job is not guaranteed to run immediately nor is that the intention of it.  It depends on the device network connection, so any custom URL job should be able to resume/run after hours or days of no network connectivity.  The custom URL job runs in the [job queue](https://github.com/yigit/android-priority-jobqueue)

* I have to cope with the application being killed and restarted arbitrarily by Android OS, so need to make sure the client can survive.  I don't believe this will be a problem. 

* I would like to see what memory difference the use of the static client will make - since at the end of the custom URL job the object is destroyed and I expect that Android's GC should come around and pick it up.

* I also want to strongly point out that latency and connection speed is not the main concern with the custom URL job - it's more important that there is stability.

**Comment by mendhak** [comment]
Also can you share the code that you tried it would be good to look at it. 

And I'll ignore the shutdown aspect since as the docs say, there will be cleanup automatically.

**Comment by updatede** [comment]
1. the job mode untouched.  job can be resume/run after hours or days of no network connectivity. okhttp will track idle timeout, server has its own idle timeout too, so connection will be closed after idle timeout. anytime a new log request come in(even after days), if there is connection, okhttp use it, if not, start a new conneciton. but instance of okhttpclient and all its data structure remains one. 
2. okhttpclient can survive, because it's a member variable, just like others , besides it's static, means only one instance for all instance of class. that is how java work.
3. okhttpclient will not be destroyed when cutstomurljob objects being destroyed, because it's static. just like your static variable "Logger LOG". it will be destroyed when reference counts is zero, which probably means ending of app.  but when new cutstomurljob objects being created, it will not be created again if been created before, also because it's static. Since this one and the only static instance of okhttpclient is a must as said in okhttp doc, so connection closed while timeout, reopened while request arrives and there's no connection, but object remains one, not one for every curstomurljob, that's exactly what we need.
4. In this scenario, low latency, fast speed, less resource consuming is not against stability, on the contrary, they can be achieved at the same time if we make a shared okhttpclient just like what okhttp doc required.
5. I use version 3.11.0 to keep track with latest okhttp for imminent tls 1.3 support.  so sslSocketFactory(SSLSocketFactory sslSocketFactory, X509TrustManager trustManager) should be used instead of  sslSocketFactory(SSLSocketFactory sslSocketFactory). but in okhttp online doc it said: "Most applications should not call this method, and instead use the system defaults. ", and now, I can't find any reason why gpslogger need customize this, so I use system defautls.
6. Since custom url job runs  one at a time, I use single static requestbuilder to customize request. I modify only customurljob.java file and use latest android studio to build. here is the code:
```java
/*
 * Copyright (C) 2016 mendhak
 *
 * This file is part of GPSLogger for Android.
 *
 * GPSLogger for Android is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * GPSLogger for Android is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with GPSLogger for Android.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.mendhak.gpslogger.loggers.customurl;


import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.birbit.android.jobqueue.Job;
import com.birbit.android.jobqueue.Params;
import com.birbit.android.jobqueue.RetryConstraint;
import com.mendhak.gpslogger.common.AppSettings;
import com.mendhak.gpslogger.common.network.Networks;
import com.mendhak.gpslogger.common.Strings;
import com.mendhak.gpslogger.common.events.UploadEvents;
import com.mendhak.gpslogger.common.slf4j.Logs;
import de.greenrobot.event.EventBus;
import okhttp3.*;

import org.slf4j.Logger;
import java.io.IOException;
import java.util.Map;


public class CustomUrlJob extends Job {

    private static final Logger LOG = Logs.of(CustomUrlJob.class);

    private static final OkHttpClient okClient = new OkHttpClient();
    private static Request.Builder requestBuilder = new Request.Builder();
    private UploadEvents.BaseUploadEvent callbackEvent;
    private CustomUrlRequest urlRequest;

    public CustomUrlJob(CustomUrlRequest urlRequest, UploadEvents.BaseUploadEvent callbackEvent) {
        super(new Params(1).requireNetwork().persist());

        this.callbackEvent = callbackEvent;
        this.urlRequest = urlRequest;
    }

    @Override
    public void onAdded() {
    }

    @Override
    public void onRun() throws Throwable {

        LOG.info("HTTP Request - " + urlRequest.getLogURL());

        requestBuilder.url(urlRequest.getLogURL());

        for(Map.Entry<String,String> header : urlRequest.getHttpHeaders().entrySet()){
            requestBuilder.header(header.getKey(), header.getValue());
        }

        if ( ! urlRequest.getHttpMethod().equalsIgnoreCase("GET")) {
            RequestBody body = RequestBody.create(null, urlRequest.getHttpBody());
            requestBuilder.method(urlRequest.getHttpMethod(),body);
        }

        Response response = okClient.newCall(requestBuilder.build()).execute();

        if (response.isSuccessful()) {
            LOG.debug("Success - response code " + response);
            EventBus.getDefault().post(callbackEvent.succeeded());
        }
        else {
            LOG.error("Unexpected response code " + response );
            EventBus.getDefault().post(callbackEvent.failed("Unexpected code " + response,new Throwable(response.body().string())));
        }

        response.body().close();
    }

    @Override
    protected void onCancel(int cancelReason, @Nullable Throwable throwable) {
        EventBus.getDefault().post(callbackEvent.failed("Could not send to custom URL", throwable));
        LOG.error("Could not send to custom URL", throwable);
    }

    @Override
    protected RetryConstraint shouldReRunOnThrowable(@NonNull Throwable throwable, int runCount, int maxRunCount) {
        return RetryConstraint.createExponentialBackoff(runCount, 3000);
    }


    @Override
    protected int getRetryLimit() {
        return 2;
    }
}
```

**Comment by updatede** [comment]
1. the job mode untouched.  job can be resume/run after hours or days of no network connectivity. okhttp will track idle timeout, server has its own idle timeout too, so connection will be closed after idle timeout. anytime a new log request come in(even after days), if there is connection, okhttp use it, if not, start a new conneciton. but instance of okhttpclient and all its data structure remains one. 
2. okhttpclient can survive, because it's a member variable, just like others , besides it's static, means only one instance for all instance of class. that is how java work.
3. okhttpclient will not be destroyed when cutstomurljob objects being destroyed, because it's static. just like your static variable "Logger LOG". it will be destroyed when reference counts is zero, which probably means ending of app.  but when new cutstomurljob objects being created, it will not be created again if been created before, also because it's static. Since this one and the only static instance of okhttpclient is a must as said in okhttp doc, so connection closed while timeout, reopened while request arrives and there's no connection, but object remains one, not one for every curstomurljob, that's exactly what we need.
4. In this scenario, low latency, fast speed, less resource consuming is not against stability, on the contrary, they can be achieved at the same time if we make a shared okhttpclient just like what okhttp doc required.
5. I use version 3.11.0 to keep track with latest okhttp for imminent tls 1.3 support.  so sslSocketFactory(SSLSocketFactory sslSocketFactory, X509TrustManager trustManager) should be used instead of  sslSocketFactory(SSLSocketFactory sslSocketFactory). but in okhttp online doc it said: "Most applications should not call this method, and instead use the system defaults. ", and now, I can't find any reason why gpslogger need customize this, so I use system defautls.
6. Since custom url job runs  one at a time, I use single static requestbuilder to customize request. I modify only customurljob.java file and use latest android studio to build. here is the code:
```java
/*
 * Copyright (C) 2016 mendhak
 *
 * This file is part of GPSLogger for Android.
 *
 * GPSLogger for Android is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * GPSLogger for Android is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with GPSLogger for Android.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.mendhak.gpslogger.loggers.customurl;


import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.birbit.android.jobqueue.Job;
import com.birbit.android.jobqueue.Params;
import com.birbit.android.jobqueue.RetryConstraint;
import com.mendhak.gpslogger.common.AppSettings;
import com.mendhak.gpslogger.common.network.Networks;
import com.mendhak.gpslogger.common.Strings;
import com.mendhak.gpslogger.common.events.UploadEvents;
import com.mendhak.gpslogger.common.slf4j.Logs;
import de.greenrobot.event.EventBus;
import okhttp3.*;

import org.slf4j.Logger;
import java.io.IOException;
import java.util.Map;


public class CustomUrlJob extends Job {

    private static final Logger LOG = Logs.of(CustomUrlJob.class);

    private static final OkHttpClient okClient = new OkHttpClient();
    private static Request.Builder requestBuilder = new Request.Builder();
    private UploadEvents.BaseUploadEvent callbackEvent;
    private CustomUrlRequest urlRequest;

    public CustomUrlJob(CustomUrlRequest urlRequest, UploadEvents.BaseUploadEvent callbackEvent) {
        super(new Params(1).requireNetwork().persist());

        this.callbackEvent = callbackEvent;
        this.urlRequest = urlRequest;
    }

    @Override
    public void onAdded() {
    }

    @Override
    public void onRun() throws Throwable {

        LOG.info("HTTP Request - " + urlRequest.getLogURL());

        requestBuilder.url(urlRequest.getLogURL());

        for(Map.Entry<String,String> header : urlRequest.getHttpHeaders().entrySet()){
            requestBuilder.header(header.getKey(), header.getValue());
        }

        if ( ! urlRequest.getHttpMethod().equalsIgnoreCase("GET")) {
           requestBuilder.method(urlRequest.getHttpMethod(),RequestBody.create(null, urlRequest.getHttpBody()));
        }

        Response response = okClient.newCall(requestBuilder.build()).execute();

        if (response.isSuccessful()) {
            LOG.debug("Success - response code " + response);
            EventBus.getDefault().post(callbackEvent.succeeded());
        }
        else {
            LOG.error("Unexpected response code " + response );
            EventBus.getDefault().post(callbackEvent.failed("Unexpected code " + response,new Throwable(response.body().string())));
        }

        response.body().close();
    }

    @Override
    protected void onCancel(int cancelReason, @Nullable Throwable throwable) {
        EventBus.getDefault().post(callbackEvent.failed("Could not send to custom URL", throwable));
        LOG.error("Could not send to custom URL", throwable);
    }

    @Override
    protected RetryConstraint shouldReRunOnThrowable(@NonNull Throwable throwable, int runCount, int maxRunCount) {
        return RetryConstraint.createExponentialBackoff(runCount, 3000);
    }


    @Override
    protected int getRetryLimit() {
        return 2;
    }
}
```

**Comment by mendhak** [comment]
The sslSocketFactory() line is for users' custom CAs and self signed certificates.  It can't be removed.  Is there a way to instantiate the client with the sslSocketFactory.  It's OK if the 1-arg method is being deprecated, the trustmanager can be obtained from `Networks.getTrustManager` just change its type to return LocalX509TrustManager.

**Comment by updatede** [comment]
But users' custom CAs and self signed certificates can be installed and trusted to system. no need to do this at application level. I usally install self signed CA to system trusted root store before I get a letsencrypt certifcate.

**Comment by mendhak** [comment]
Ah no can't do that :(   It's done for distribution and convenience - many organisations distribute GPSLogger this way - the app and keystore are bundled together since their users aren't savvy enough to add it themselves.  And the convenience aspect is for users to add their own certs in or just move the file about as they'd like. 

But it must be possible - haven't tried this yet, but I could try to initialize it in a static constructor?

**Comment by updatede** [comment]
I modify two places to get sslSocketFactory compiled, it seems good for me, here it is:
```java
     public static X509TrustManager getTrustManager(Context context)
            throws CertificateException, NoSuchAlgorithmException, KeyStoreException, IOException, CertStoreException {
        return new LocalX509TrustManager(getKnownServersStore(context));
     }
```

```java
    private static final OkHttpClient okClient ;
    static {
        try {
            okClient = new OkHttpClient.Builder().sslSocketFactory(Networks.getSocketFactory(AppSettings.getInstance()),Networks.getTrustManager(AppSettings.getInstance())).build();
        } catch (Exception e) {
            throw new Error(e);
        }
    }
    private static Request.Builder requestBuilder = new Request.Builder();
```
About distribution and convenience , I think it will be good if user can import all preference settings. I know new version can share old version's settings, but for newly installed gpslogger user must manually input settings on every mobile device.

**Comment by updatede** [comment]
it's more simple to use deprecated sslSocketFactory, need not to modify `Networks.getTrustManager` and handle exception in static initialization,
```java
private static final OkHttpClient okClient = new OkHttpClient.Builder().sslSocketFactory(Networks.getSocketFactory(AppSettings.getInstance())).build();
private static Request.Builder requestBuilder = new Request.Builder();
```

**Comment by mendhak** [comment]
I did some testing finally and was able to get some nice representation of the difference.   It took a while as I moved house in between and it takes time to unpack and get back into things. 

Here is the before.   Notice the CPU spikes and larger network activity whenever the client is used.  Memory usage varied from 90 to 200 MB.

![image](https://user-images.githubusercontent.com/746276/49830846-3f0e0e00-fd8a-11e8-8672-191ff8fbef21.png)

Here is the after with your changes.  Tinier CPU spikes and smaller network activity.  Memory usage was also lower at about 90 MB.

![image](https://user-images.githubusercontent.com/746276/49830889-5d740980-fd8a-11e8-9fa9-1f802aaa5cb6.png)

That's really good, many thanks.  I've now merged this into milestone98 branch.
